src/docs_pseudo_methods.cr

```diff
@@ -0,0 +1,134 @@
+  # *type* must be a constant, it cannot be evaluated at runtime.
+  #
+  # ```
+  # a = 1
+  # a.class                 # => Int32
+  # a.is_a?(Int32)          # => true
+  # a.is_a?(String)         # => false
+  # a.is_a?(Number)         # => true
+  # a.is_a?(Int32 | String) # => true
+  # ```
+  #
+  # NOTE: This is a pseudo-method provided directly by the Crystal compiler.
+  # It cannot be redefined nor overridden.
+  def __crystal_pseudo_is_a?(type : Class) : Bool
+  end
+
+  # Returns `true` if `self` is `Nil`.
+  #
+  # ```
+  # 1.nil?   # => false
+  # nil.nil? # => true
+  # ```
+  #
+  # This method is equivalent to `is_a?(Nil)`.
+  #
+  # NOTE: This is a pseudo-method provided directly by the Crystal compiler.
+  # It cannot be redefined nor overridden.
+  def __crystal_pseudo_nil? : Bool
+  end
+
+  # Returns `self`.
+  #
+  # The type of this expression is restricted to *type* by the compiler.
+  # *type* must be a constant, it cannot be evaluated at runtime.
+  #
+  # If *type* is not a valid restriction for the expression type, it
+  # is a compile-time error.
+  # If *type*  is a valid restriction for the expression, but `self` can't
+  # be restricted to *type*, it raises at runtime.
+  # *type* may be a wider restriction than the expression type, the resulting
+  # type is narrowed to the minimal restriction.
+  #
+  # ```
+  # a = [1, "foo"][0]
+  # typeof(a)             # => Int32 | String
+  #
+  # typeof(a.as(Int32))   # => Int32
+  # a.as(Int32)           # => 1
+  #
+  # typeof(a.as(Bool))    # Compile Error: can't cast (Int32 | String) to Bool
+  #
+  # typeof(a.as(String))  # => String
+  # a.as(String)          # Runtime Error: cast from Int32 to String failed
+  #
+  # typeof(a.as(Int32 | Bool)) # => Int32
+  # a.as(Int32 | Bool)         # => 1
+  # ```
+  #
+  # NOTE: This is a pseudo-method provided directly by the Crystal compiler.
+  # It cannot be redefined nor overridden.
+  def __crystal_pseudo_as(type : Class)
+  end
+
+  # Returns `self` or `nil` if can't be restricted to *type*.
+  #
+  # The type of this expression is restricted to *type* by the compiler.
+  # If *type* is not a valid type restriction for the expression type, then
+  # it is restricted to `Nil`.
+  # *type* must be a constant, it cannot be evaluated at runtime.
+  #
+  # ```
+  # a = [1, "foo"][0]
+  # typeof(a)              # => Int32 | String
+  #
+  # typeof(a.as?(Int32))   # => Int32 | Nil
+  # a.as?(Int32)           # => 1
+  #
+  # typeof(a.as?(Bool))    # => Nil
+  # a.as?(Bool)            # => nil
+  #
+  # typeof(a.as?(String))  # => String | Nil
+  # a.as?(String)          # nil
+  #
+  # typeof(a.as?(Int32 | Bool)) # => Int32 | Nil
+  # a.as?(Int32 | Bool)         # => 1
+  # ```
+  #
+  # NOTE: This is a pseudo-method provided directly by the Crystal compiler.
+  # It cannot be redefined nor overridden.
+  def __crystal_pseudo_as?(type : Class)
+  end
+
+  # Returns `true` if `self` responds to *name*.
```

Comment:
It might help more to say something along the lines:
``Returns `true` if the method *name* can be called on `self` ``